### ISSUE TITLE
JSpecify: false positive when a method returns `? extends T` and `T` has a `@Nullable` upper bound (support `NO_CHANGE` nullness operator better)

### ISSUE THREAD

**Issue by vlsi** [uber/NullAway#1727]
### Version

NullAway 0.14.0, and current `master` (21f02dd). Error Prone 2.50.0, JDK 21.

Flags: `-XepOpt:NullAway:JSpecifyMode=true -XepOpt:NullAway:HandleWildcardGenerics=true -XDaddTypeAnnotationsToSymbol=true`. `HandleWildcardGenerics` defaults to `JSpecifyExperimental`, so `JSpecifyExperimental=true` triggers it too. With wildcard handling off, nothing is reported.

### Summary

A method whose return type is a type variable `R` instantiated to `? extends T` produces a `@Nullable` result whenever `T` is declared `T extends @Nullable Object`, even though the wildcard's bound is `T` itself. Returning that result as `T` is then reported, and so is passing it to anything that expects `T`.

The wildcard's bound is what gets lost: `Box<? extends String>` and `Box<T>` are both handled correctly, and `? extends T` misbehaves only when `T` has a `@Nullable` upper bound. That suggests NullAway falls back to the *type parameter's* declared bound (`@Nullable Object` on `Box`'s `R`) instead of reading the wildcard's own bound.

### Reproducer

```java
import org.jspecify.annotations.NullMarked;
import org.jspecify.annotations.Nullable;

@NullMarked
public class Test {
  interface Box<R extends @Nullable Object> {
    R get();
  }

  static <T extends @Nullable Object> T unwrap(Box<? extends T> b) {
    // BUG: Diagnostic contains: returning @Nullable expression from method with @NonNull return type
    return b.get();
  }
}
```

```
Test.java:11: warning: [NullAway] returning @Nullable expression from method with @NonNull return type
    return b.get();
    ^
```

Note that the message calls `T` a `@NonNull` return type although `T` is bounded by `@Nullable Object`.

### What does and does not report

Same `Box`, same flags:

| Signature | Reported |
| --- | --- |
| `<T extends @Nullable Object> T unwrap(Box<T> b)` | no |
| `<T> T unwrap(Box<? extends T> b)` | no |
| `String unwrap(Box<? extends String> b)` | no |
| `<T extends @Nullable Object> Box<T> unwrap(Box<? extends Box<T>> b)` | no |
| `<T extends @Nullable Object> T unwrap(Box<? extends T> b)` | **yes** |

The result also poisons a local. Storing it in a `T` local reports nothing on its own, but every later use of that local that requires non-null is reported:

```java
static <T extends @Nullable Object> void pass(Box<? extends T> b) {
  T t = b.get();
  // BUG: Diagnostic contains: passing @Nullable parameter 't' where @NonNull is required
  Test.<T>sink(t);
}

static <T extends @Nullable Object> void sink(T t) {}
```

### Why the wildcard is the right declaration

The wildcard is not decoration. In a reduce whose accumulator starts at null:

```java
static <S, A extends @Nullable Object> A aggregate(
    Iterable<S> source, A seed, Fn2<A, S, ? extends A> func) { ... }
```

callers pass reducers that accept a null accumulator but never produce one, so their type is `Fn2<@Nullable Integer, Integer, Integer>`. Without the wildcard, generic invariance rejects them against `Fn2<A, S, A>` where `A` is `@Nullable Integer`.

### Workaround

`@SuppressWarnings("NullAway")` on the enclosing method.

### Where we hit it

`EnumerableDefaults.aggregate`, while replacing the Checker Framework with NullAway in Apache Calcite (https://github.com/apache/calcite/pull/5213).

**Comment by vlsi** [comment]
Another sink for the same wrong nullness: assigning the result to a field of type `T`. Verified on `master` (21f02dd) with the same flags.

```java
import java.util.Iterator;
import org.jspecify.annotations.NullMarked;
import org.jspecify.annotations.Nullable;

@NullMarked
public class Test {
  interface Box<R extends @Nullable Object> {
    R get();
  }

  static class Cursor<T extends @Nullable Object> {
    private final Box<? extends T> box;
    T current;

    // BUG: Diagnostic contains: assigning @Nullable expression to @NonNull field
    Cursor(Box<? extends T> box) {
      this.box = box;
      this.current = box.get();
    }

    // BUG: Diagnostic contains: assigning @Nullable expression to @NonNull field
    void advance() {
      current = box.get();
    }
  }
}
```

The constructor assignment also drags a second, cascading report along with it:

```
Test.java:17: warning: [NullAway] initializer method does not guarantee @NonNull field 'current' (line 14) is initialized along all control-flow paths (remember to check for exceptions or early returns).
```

The field variant splits the same way as the return variant: `Box<T>` is clean, `<T> ... Box<? extends T>` with a non-null bound is clean, and only `<T extends @Nullable Object> ... Box<? extends T>` reports.

A JDK type behaves identically, so library models are not involved:

```java
static class Cursor<T extends @Nullable Object> {
  private final Iterator<? extends T> iterator;
  T current;

  Cursor(Iterator<? extends T> iterator) {
    this.iterator = iterator;
    // BUG: Diagnostic contains: assigning @Nullable expression to @NonNull field
    this.current = iterator.next();
  }
}
```

Found in the same Calcite migration (https://github.com/apache/calcite/pull/5213).

**Comment by msridhar** [comment]
@vlsi I just wanted to thank you quickly for all these great reports and for testing JSpecifyExperimental!  It'll take us some time to go through the reports, particularly as I'm juggling a few other things.  But, we'll get to them as soon as we can, and try to fix as much as possible for the next release.

**Comment by msridhar** [comment]
This one is going to require some work, since it requires better modeling the [`NO_CHANGE` nullness operator](https://jspecify.dev/docs/spec/#nullness-operator) from the JSpecify spec, rather than just treating expressions as being either `@Nullable` or `@NonNull`.  When we have code like:
```java
static <T extends @Nullable Object> T unwrap(Box<? extends T> b) {
  // BUG: Diagnostic contains: returning @Nullable expression from method with @NonNull return type
  return b.get();
}
```
We want to reason that the nullability of `b.get()` will always be compatible with the nullness of `T`, so the code is safe.  But, if we have code like:
```java
static class Foo<T extends @Nullable Object> {
  T get() { throw new RuntimeException(); }
}
static class NullableBound<U extends @Nullable Object> {
  void test(Foo<? extends U> f) {
    // BUG: Diagnostic contains: dereferenced expression 'f.get()' is @Nullable
    f.get().hashCode();
  }
}
```
We want to reason that `f.get()` _could_ be `@Nullable`, so we should warn on a dereference ([relevant spec section](https://jspecify.dev/docs/spec/#null-exclusive-under-every-parameterization)).

@vlsi how critical is this one for you all?  For now I plan to de-prioritize to see if there are easier issues we can quickly fix first.

**Comment by vlsi** [comment]
> how critical is this one for you all?

Looks like this is not that critical.
I've completed all `main/java` Calcite migration to NullAway, and this issue impacts 3 `@SuppressWarnings("NullAway")` cases only.
It is definitely workable.

PS. I've spent some time with 5.6 Sol and Opus 5, and they agree the change is not trivial. I wonder if you want to fix it on your own, or if you are fine if I protopye A-B-C split (see `open questions` in https://github.com/uber/NullAway/pull/1767). In any case I've filed a PR with tests which I am sure won't hurt.

**Comment by msridhar** [comment]
Thanks a lot for the tests @vlsi!  I probably want to think about design a bit on this one.  The change will end up touching some core generics-checking code, and I want to make sure I keep the design clean when we do it.  I might even decide I need to hand code some of it 🙂, or refactor the current code to make it easier.  But, before getting into this one, I'd like to get the current `JSpecifyExperimental` features to be more solid, with the goal of enabling those by default.

By the way, the issues here go deeper than wildcards, as you might have noticed.  E.g., this code currently passes NullAway with no warning, though there should be one:
```java
@NullMarked
class Test {
  static class Foo<T extends @Nullable Object> {
      int test(T t) {
         return t.hashCode();  // there should be a warning here!
      }
  }
}
```
We may want to get things working on tests that don't involve wildcards first.  If we code the subtyping and dereference checks cleanly, hopefully much wildcard-related functionality will reuse that logic and work cleanly after that.
